src/esp/geo/geo.cpp

```diff
@@ -58,23 +61,66 @@ std::vector<vec2f> convexHull2D(const std::vector<vec2f>& points) {
 
   return hull;
 }
+/**
+ * Assume the aabb is expressed as the center 'c' and the extent 'e'.
+ * each corner X is nothing but a combination from
+ * (c_x +/- e_x, c_y +/- e_y, c_z +/- e_z)
+ *
+ * Denote y = (+/- e_x, +/- e_y, +/- e_z)
+ *
+ * The corner is transformed by:
+ * x = R * x0 + t,
+ * where x_0, x are the coordinates before and after transformation, t is the
+ * translation.
+ *
+ * x = R * (c0 + y0) + t  = (Rc0 + t) + Ry0    eq(1)
+ *
+ * Our Goal is to find the x_max and x_min after the transformation.
+ *
+ * First, determing the x_max:
+ *
+ * In eq(1), Rc0 + t is a constant, which means max{x} iff max{Ry0}
+ *
+ * R looks like:
+ * [R0, R1, R2]
+ * [R3, R4, R5]
+ * [R6, R7, R8]
+ *
+ * We focus on the 1st entry (the 'x' entry) of Ry0:
+ * y_x =<(R0, R1, R2), (+/- e_x, +/- e_y, +/- e_z)>
+ *     =<(+/- R0, +/- R1, +/- R2), (e_x, e_y, e_z)>
+ *
+ * Now, note that e_x, e_y, e_z are all positive values for any non-degenerate
+ * aabb.
+ *
+ * So y_x reaches MAX when +/- R0, +/- R1, +/- R2 are all >=0
+ * that means max{y_x} = <(|R0|, |R1|, |R2|), (e_x, e_y, e_z)>
+ * (|R0| means the absolute value of R0)
+ *
+ * and likewise for y_y, y_z
+ *
+ * The derivation for x_min is similar since same logic applies.
+ */
+Mn::Range3D getTransformedBB(const Mn::Range3D& range,
+                             const Mn::Matrix4& xform) {
+  Mn::Matrix3x3 absRotationScaling = xform.rotationScaling();
+  // compute the absolute value of the rotationScaling part of the original
+  // transformation matrix
+  for (int iRow = 0; iRow < 3; ++iRow) {
+    for (int jCol = 0; jCol < 3; ++jCol) {
+      absRotationScaling[iRow][jCol] =
+          std::fabs(absRotationScaling[iRow][jCol]);
+    }
+  }
+  const Mn::Vector3 center = (range.min() + range.max()) / 2.0;
+  const Mn::Vector3 extent = (range.max() - range.min()) / 2.0;
```

Comment:
```suggestion
  const Mn::Vector3 extent = range.size()/2.0f;
```

Hmm, I think I could add an `extent()` accessor there too... oh and probably related: for some `Math::Intersection` routines we went with having a special overload that takes center + extent directly, to save the subtraction and division from the above calculation. I planto add a variant of the `Range` type that stores center + extent instead of min/max to have a builtin type for these use cases, and could prioritize that if you think storing everything as center + extent here would speed things up.

Reply:
Interesting. Let me think which one is preferable. <min, max> or <center, extent>

Reply:
Okay, adding to my TODO list :)